src/pieces.py

```diff
@@ -9,118 +9,272 @@
+    """A King.
+
+    Moves one square in any direction that is not blocked by same side piece or targeted by enemy piece.
+    """
+
+    steps = Melee.steps.union(Piece.straights.union(Piece.diagonals))
+
+    def __repr__(self) -> str:
+        super().__repr__.__doc__
+        return "♚" if self.is_black else "♔"
 
 
 @dataclass
-class Knight(Piece):
-    """A knight."""
+class Knight(Melee):
+    """A knight.
+
+    Moves two squares in any direction and one square in a vertical to that direction as one move.
+    It therefore can jump over other pieces. Target square must not be blocked by a same side piece.
+    """
+
+#   Knight moves
+    steps = Melee.steps.union(diagonal + straight for diagonal, straight in product(Piece.diagonals, Piece.straights))
+    steps = steps.difference(Piece.rizontals.union(Piece.straights))  # remove the non-knight moves
 
     value: int = 3
 
     def __repr__(self) -> str:
-        """Represent the knight.
-
-        Returns:
-            The representation of a knight.
-        """
+        super().__repr__.__doc__
         return "♞" if self.is_black else "♘"
 
 
 @dataclass
-class Rook(Piece):
-    """A rook."""
+class Range(Piece):
+    """A long range piece:
+        - Rook
+        - Bishop
+        - Queen
+    """
+
+    def legal_moves(self, square: Square, condition: Callable[[Square], bool]) -> set[Square]:
+        super().legal_moves.__doc__
+        squares = set()
+        for move in self.steps:  # For each direction.
+            advanced_square = square + move  # Start looking at advanced squares
+            while condition(advanced_square):  # While we don't hit something.
+                squares.add(advanced_square)  # Add the damn square to legal destination squares.
+                advanced_square += move  # Advance to the next square in line.
+        return squares
+
+
+@dataclass
+class Rook(Range):
+    """A rook.
+
+    Moves anywhere along its rank xor its file, unless blocked by enemy (which captures) or same piece.
+    """
+
+#   Straight lines:
+    steps = Range.steps.union(Piece.straights)
 
     value: int = 5
 
     def __repr__(self) -> str:
-        """Represent the rook.
-
-        Returns:
-            The representation of a rook.
-        """
+        super().__repr__.__doc__
         return "♜" if self.is_black else "♖"
 
 
 @dataclass
-class Queen(Piece):
-    """A queen."""
+class Bishop(Range):
+    """A bishop.
 
-    value: int = 9
+    Moves anywhere along a diagonal, unless blocked by enemy (which captures) or same piece.
+    This means that a bishop never leaves the color of the square it is on.
+    A player starts with two bishops:
+        A black-square bishop.
+        A white-square bishop.
+    So both cover the squares of the whole board, and it is the reason the worth more when both on the board.
+    """
 
-    def __repr__(self) -> str:
-        """Represent the queen.
+#   Diagonal lines:
+    steps = Range.steps.union(Piece.diagonals)
```

Comment:
```py
steps = set(Piece.diagonals)
```
(To create a separate object than `Piece.diagonals`)

Reply:
So at first, I thought the set steps would gradually grow as we specialize more, but in the end implementation, it was unnecessary. I think I will adopt this approach.